File: src/main/java/org/example/bankramenserver/domain/recurring/service/DetectRecurringPaymentService.java
package org.example.bankramenserver.domain.recurring.service;

import lombok.RequiredArgsConstructor;
import org.example.bankramenserver.domain.push.domain.PushNotification;
import org.example.bankramenserver.domain.push.service.SendPushNotificationService;
import org.example.bankramenserver.domain.recurring.domain.RecurringPayment;
import org.example.bankramenserver.domain.recurring.domain.repository.RecurringPaymentRepository;
import org.example.bankramenserver.domain.transaction.domain.Transaction;
import org.example.bankramenserver.domain.transaction.domain.repository.TransactionRepository;
import org.example.bankramenserver.domain.user.domain.User;
import org.example.bankramenserver.domain.user.domain.repository.UserRepository;
import org.example.bankramenserver.global.util.MoneyFormatter;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Map;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class DetectRecurringPaymentService {

    private static final int MONTHLY_TOLERANCE_DAYS = 3;
    private static final int YEARLY_TOLERANCE_DAYS = 7;

    private final TransactionRepository transactionRepository;
    private final RecurringPaymentRepository recurringPaymentRepository;
    private final UserRepository userRepository;
    private final SendPushNotificationService sendPushNotificationService;

    @Transactional
    public void execute(UUID transactionId) {
        Transaction current = transactionRepository.findById(transactionId)
                .orElseThrow();

        if (current.getType() != Transaction.TransactionType.EXPENSE) return;
        if (current.getDescription() == null || current.getDescription().isBlank()) return;

        detectMonthly(current);
        detectYearly(current);
    }

    private void detectMonthly(Transaction current) {
        LocalDate expectedPreviousDate = current.getTransactionDate().minusMonths(1);

        boolean exists = transactionRepository.existsSameExpenseTransactionBetween(
                current.getUser().getId(),
                current.getDescription(),
                current.getAmount(),
                expectedPreviousDate.minusDays(MONTHLY_TOLERANCE_DAYS),
                expectedPreviousDate.plusDays(MONTHLY_TOLERANCE_DAYS)
        );

        if (exists) {
            createCandidateIfNotExists(current, RecurringPayment.Cycle.MONTHLY);
        }
    }

    private void detectYearly(Transaction current) {
        LocalDate expectedPreviousDate = current.getTransactionDate().minusYears(1);

        boolean exists = transactionRepository.existsSameExpenseTransactionBetween(
                current.getUser().getId(),
                current.getDescription(),
                current.getAmount(),
                expectedPreviousDate.minusDays(YEARLY_TOLERANCE_DAYS),
                expectedPreviousDate.plusDays(YEARLY_TOLERANCE_DAYS)
        );

        if (exists) {
            createCandidateIfNotExists(current, RecurringPayment.Cycle.YEARLY);
        }
    }

    private void createCandidateIfNotExists(Transaction current, RecurringPayment.Cycle cycle) {

        boolean alreadyExists = recurringPaymentRepository
                .existsByUser_IdAndNameAndAmountAndCycleAndActiveTrue(
                        current.getUser().getId(),
                        current.getDescription(),
                        current.getAmount(),
                        cycle
                );

        if (alreadyExists) return;

        User user = userRepository.getReferenceById(current.getUser().getId());

        RecurringPayment recurringPayment = RecurringPayment.builder()
                .user(user)
                .category(current.getCategory())
                .name(current.getDescription())
                .amount(current.getAmount())
                .cycle(cycle)
                .billingDay(current.getTransactionDate().getDayOfMonth())
                .nextBillingDate(calculateNextBillingDate(current.getTransactionDate(), cycle))
                .registrationType(RecurringPayment.RegistrationType.AUTO_DETECTED)
                .confirmed(false)
                .build();

        recurringPaymentRepository.save(recurringPayment);

        sendPushNotificationService.execute(
                current.getUser().getId(),
                PushNotification.NotificationType.RECURRING_CANDIDATE,
                "정기결제 같아요",
                "%s %s원이 반복 결제되고 있어요. 정기결제로 등록할까요?"
                        .formatted(
                                current.getDescription(),
                                MoneyFormatter.format(current.getAmount())
                        ),
                recurringPayment.getId().toString(),
                Map.of(
                        "type", PushNotification.NotificationType.RECURRING_CANDIDATE.name(),
                        "recurringPaymentId", recurringPayment.getId().toString()
                )
        );
    }

    private LocalDateTime calculateNextBillingDate(LocalDate transactionDate, RecurringPayment.Cycle cycle) {
        if (cycle == RecurringPayment.Cycle.MONTHLY) {
            return transactionDate.plusMonths(1).atStartOfDay();
        }
        return transactionDate.plusYears(1).atStartOfDay();
    }
}

File: src/main/kotlin/org/example/balogserver/domain/recurring/service/DetectRecurringPaymentService.kt
package org.example.balogserver.domain.recurring.service

import org.example.balogserver.domain.push.domain.PushNotification
import org.example.balogserver.domain.push.service.SendPushNotificationService
import org.example.balogserver.domain.recurring.domain.RecurringPayment
import org.example.balogserver.domain.recurring.domain.repository.RecurringPaymentRepository
import org.example.balogserver.domain.transaction.domain.Transaction
import org.example.balogserver.domain.transaction.domain.repository.TransactionRepository
import org.example.balogserver.domain.user.domain.repository.UserRepository
import org.example.balogserver.global.util.MoneyFormatter
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.UUID

@Service
class DetectRecurringPaymentService(
    private val transactionRepository: TransactionRepository,
    private val recurringPaymentRepository: RecurringPaymentRepository,
    private val userRepository: UserRepository,
    private val sendPushNotificationService: SendPushNotificationService,
) {
    @Transactional
    fun execute(transactionId: UUID) {
        val current = transactionRepository.findById(transactionId).orElseThrow()
        if (current.type != Transaction.TransactionType.EXPENSE || current.description.isNullOrBlank()) return
        detect(current, RecurringPayment.Cycle.MONTHLY, current.transactionDate.minusMonths(1), 3)
        detect(current, RecurringPayment.Cycle.YEARLY, current.transactionDate.minusYears(1), 7)
    }

    private fun detect(current: Transaction, cycle: RecurringPayment.Cycle, expectedDate: LocalDate, toleranceDays: Long) {
        if (transactionRepository.existsSameExpenseTransactionBetween(current.user.id, current.description, current.amount, expectedDate.minusDays(toleranceDays), expectedDate.plusDays(toleranceDays))) createCandidateIfNotExists(current, cycle)
    }

    private fun createCandidateIfNotExists(current: Transaction, cycle: RecurringPayment.Cycle) {
        val userId = current.user.id!!
        if (recurringPaymentRepository.existsByUser_IdAndNameAndAmountAndCycleAndActiveTrue(userId, current.description, current.amount, cycle)) return
        val recurringPayment = RecurringPayment.builder().user(userRepository.getReferenceById(userId)).category(current.category).name(current.description).amount(current.amount)
            .cycle(cycle).billingDay(current.transactionDate.dayOfMonth).nextBillingDate(nextBillingDate(current.transactionDate, cycle))
            .registrationType(RecurringPayment.RegistrationType.AUTO_DETECTED).confirmed(false).build()
        recurringPaymentRepository.save(recurringPayment)
        sendPushNotificationService.execute(
            userId, PushNotification.NotificationType.RECURRING_CANDIDATE, "정기결제 같아요",
            "${current.description} ${MoneyFormatter.format(current.amount)}원이 반복 결제되고 있어요. 정기결제로 등록할까요?", recurringPayment.id.toString(),
            mapOf("type" to PushNotification.NotificationType.RECURRING_CANDIDATE.name, "recurringPaymentId" to recurringPayment.id.toString()),
        )
    }

    private fun nextBillingDate(transactionDate: LocalDate, cycle: RecurringPayment.Cycle): LocalDateTime =
        (if (cycle == RecurringPayment.Cycle.MONTHLY) transactionDate.plusMonths(1) else transactionDate.plusYears(1)).atStartOfDay()
}
